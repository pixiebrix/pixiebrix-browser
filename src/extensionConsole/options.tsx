/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import "@/vendors/theme/app/app.scss";
import "@/vendors/overrides.scss";
import "@/utils/layout.scss";
import "./options.scss";

import "@/extensionContext";
import "@/development/darkMode";

import { render } from "react-dom";
import React from "react";
import App from "@/extensionConsole/App";
import initGoogle from "@/contrib/google/initGoogle";
import { initToaster } from "@/utils/notify";
import { initTelemetry } from "@/background/messenger/api";
import registerMessenger from "@/extensionConsole/messenger/registration";
import { initPerformanceMonitoring } from "@/telemetry/performance";

function init(): void {
  render(<App />, document.querySelector("#container"));
}

registerMessenger();
void initGoogle();
initToaster();
initTelemetry();
void initPerformanceMonitoring();
init();
