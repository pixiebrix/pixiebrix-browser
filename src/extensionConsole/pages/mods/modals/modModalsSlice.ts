/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { createSlice, type PayloadAction } from "@reduxjs/toolkit";
import { type MessageContext } from "@/types/loggerTypes";
import { type RegistryId } from "@/types/registryTypes";
import { type UUID } from "@/types/stringTypes";

type LogsContext = {
  title: string;
  messageContext: MessageContext;
};

export type ShareContext = {
  blueprintId?: RegistryId;
  extensionId?: UUID;
};

export type PublishContext = ShareContext & {
  cancelingPublish?: boolean;
};

type ModModalsState = {
  showLogsContext: LogsContext;
  showShareContext: ShareContext;
  showPublishContext: PublishContext;
};

export type ModModalsRootState = {
  modModals: ModModalsState;
};

const initialState: ModModalsState = Object.freeze<ModModalsState>({
  showLogsContext: null,
  showShareContext: null,
  showPublishContext: null,
});

export const modModalsSlice = createSlice({
  name: "modModals",
  initialState,
  reducers: {
    setLogsContext(state, action: PayloadAction<LogsContext>) {
      state.showLogsContext = action.payload;
      state.showShareContext = null;
      state.showPublishContext = null;
    },
    setShareContext(state, action: PayloadAction<ShareContext | null>) {
      state.showShareContext = action.payload;
      state.showLogsContext = null;
      state.showPublishContext = null;
    },
    setPublishContext(state, action: PayloadAction<PublishContext | null>) {
      state.showPublishContext = action.payload;
      state.showLogsContext = null;
      state.showShareContext = null;
    },
    setCancelingPublish(state) {
      state.showPublishContext.cancelingPublish = true;
    },
    closeModal() {
      return initialState;
    },
  },
});
