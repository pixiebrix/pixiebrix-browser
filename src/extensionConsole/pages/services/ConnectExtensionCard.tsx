/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useState } from "react";
import { useAsyncState } from "@/hooks/common";
import { getBaseURL } from "@/services/baseService";
import { useAsyncEffect } from "use-async-effect";
import { isLinked } from "@/auth/token";
import { Card } from "react-bootstrap";
import urljoin from "url-join";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faExternalLinkAlt } from "@fortawesome/free-solid-svg-icons";

const ConnectExtensionCard: React.FunctionComponent = () => {
  const [hasExtensionKey, setHasExtensionKey] = useState(true);
  const [serviceURL] = useAsyncState(getBaseURL);

  useAsyncEffect(
    async (isMounted) => {
      const hasKey = await isLinked();
      if (isMounted()) return;
      setHasExtensionKey(hasKey);
    },
    [setHasExtensionKey]
  );

  if (hasExtensionKey) {
    return null;
  }

  return (
    <Card className="mb-4">
      <Card.Header>Connect to PixieBrix</Card.Header>
      <Card.Body>
        <p>
          By linking your browser extension with PixieBrix, you&apos;ll get
          access to team features and public services.
        </p>
        {serviceURL && (
          <a
            href={urljoin(serviceURL, "extension")}
            className="btn btn-primary"
            rel="noreferrer"
            target="_blank"
          >
            <FontAwesomeIcon icon={faExternalLinkAlt} /> Open PixieBrix Website
          </a>
        )}
      </Card.Body>
    </Card>
  );
};

export default ConnectExtensionCard;
