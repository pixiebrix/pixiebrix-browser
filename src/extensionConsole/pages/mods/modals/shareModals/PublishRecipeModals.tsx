/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { selectShowPublishContext } from "@/extensionConsole/pages/mods/modals/modModalsSelectors";
import { modModalsSlice } from "@/extensionConsole/pages/mods/modals/modModalsSlice";
import PublishRecipeContent from "./PublishRecipeContent";
import { Modal } from "react-bootstrap";
import { useGetMarketplaceListingQuery } from "@/services/api";
import Loader from "@/components/Loader";
import { useOptionalModDefinition } from "@/modDefinitions/modDefinitionHooks";
import EditPublishContent from "./EditPublishContent";
import CancelPublishContent from "./CancelPublishContent";
import PublishedContent from "./PublishedContent";

const ModalContentSwitch: React.FunctionComponent = () => {
  const showPublishContext = useSelector(selectShowPublishContext);
  const { blueprintId: modId, cancelingPublish } = showPublishContext;
  const { data: listing, isLoading: isLoadingListing } =
    useGetMarketplaceListingQuery({ packageId: modId });
  const { data: modDefinition, isLoading: isLoadingRecipe } =
    useOptionalModDefinition(modId);

  if (isLoadingRecipe || isLoadingListing) {
    return (
      <Modal.Body>
        <Loader />
      </Modal.Body>
    );
  }

  if (!modDefinition.sharing.public) {
    return <PublishRecipeContent />;
  }

  // A mod is pending publish if it has been made public but does not yet have a marketplace listing
  if (!listing) {
    return cancelingPublish ? <CancelPublishContent /> : <EditPublishContent />;
  }

  // Normally we shouldn't get here, because the Publish action is not displayed for a published blueprint
  return <PublishedContent />;
};

const PublishRecipeModals: React.FunctionComponent = () => {
  const dispatch = useDispatch();
  const closeModal = () => {
    dispatch(modModalsSlice.actions.closeModal());
  };

  const showPublishContext = useSelector(selectShowPublishContext);
  const showPublishRecipeModal = showPublishContext?.blueprintId != null;

  return (
    <Modal show={showPublishRecipeModal} onHide={closeModal}>
      {showPublishRecipeModal && <ModalContentSwitch />}
    </Modal>
  );
};

export default PublishRecipeModals;
