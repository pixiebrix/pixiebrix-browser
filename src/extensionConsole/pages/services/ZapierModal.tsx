/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// eslint-disable-next-line no-restricted-imports -- TODO: Fix over time
import { Button, Modal, Form, InputGroup } from "react-bootstrap";
import React, { useCallback } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCopy } from "@fortawesome/free-solid-svg-icons";
import notify from "@/utils/notify";
import { reportEvent } from "@/telemetry/events";
import { writeTextToClipboard } from "@/utils/clipboardUtils";
import { useGetZapierKeyQuery } from "@/services/api";

interface OwnProps {
  onClose: () => void;
}

const ZapierModal: React.FunctionComponent<OwnProps> = ({ onClose }) => {
  const { data } = useGetZapierKeyQuery();

  const handleCopy = useCallback(async () => {
    await writeTextToClipboard(String(data?.api_key));
    notify.success("Copied API Key to clipboard");
    reportEvent("ZapierKeyCopy");
  }, [data?.api_key]);

  return (
    <Modal show onHide={onClose} backdrop="static" keyboard={false}>
      <Modal.Header closeButton>
        <Modal.Title>Zapier</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <p>Use this key to log in to PixieBrix from Zapier</p>
        <Form.Group>
          <Form.Group controlId="label">
            <Form.Label>PixieBrix API Key</Form.Label>
            <InputGroup>
              <Form.Control type="text" readOnly defaultValue={data?.api_key} />
              <InputGroup.Append>
                <Button variant="info" onClick={handleCopy}>
                  <FontAwesomeIcon icon={faCopy} />
                </Button>
              </InputGroup.Append>
            </InputGroup>
          </Form.Group>
        </Form.Group>
      </Modal.Body>
    </Modal>
  );
};

export default ZapierModal;
