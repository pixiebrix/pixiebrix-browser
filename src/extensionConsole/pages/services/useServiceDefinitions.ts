/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { useSelector } from "react-redux";
import { type RootState } from "@/store/optionsStore";
import { useParams } from "react-router";
import { useMemo } from "react";
import { sortBy } from "lodash";
import registry from "@/services/registry";
import { useAsyncState } from "@/hooks/common";
import { PIXIEBRIX_INTEGRATION_ID } from "@/services/constants";
import { selectConfiguredServices } from "@/store/services/servicesSelectors";
import {
  type Integration,
  type IntegrationConfig,
} from "@/types/integrationTypes";
import { type UUID } from "@/types/stringTypes";

interface IntegrationDefinitions {
  serviceDefinitions: Integration[];
  activeConfiguration: IntegrationConfig | null;
  activeService: Integration | null;
  isPending: boolean;
  showZapier: boolean;
}

const ZAPIER_SLUG = "zapier";

function useServiceDefinitions(): IntegrationDefinitions {
  const configuredServices = useSelector<RootState, IntegrationConfig[]>(
    selectConfiguredServices
  );
  const { id: configurationId } = useParams<{ id: UUID }>();

  const showZapier = configurationId === ZAPIER_SLUG;

  const [serviceDefinitions, isPending, error] = useAsyncState(
    async () =>
      sortBy(
        // eslint-disable-next-line unicorn/no-await-expression-member
        (await registry.all()).filter((x) => x.id !== PIXIEBRIX_INTEGRATION_ID),
        (x) => x.id
      ),
    []
  );

  const activeConfiguration = useMemo(
    () =>
      configurationId && configurationId !== ZAPIER_SLUG
        ? configuredServices.find((x) => x.id === configurationId)
        : null,
    [configuredServices, configurationId]
  );

  const activeService = useMemo(
    () =>
      activeConfiguration
        ? (serviceDefinitions ?? []).find(
            (x) => x.id === activeConfiguration.serviceId
          )
        : null,
    [serviceDefinitions, activeConfiguration]
  );

  if (error) {
    throw error;
  }

  return {
    activeConfiguration,
    serviceDefinitions,
    activeService,
    isPending,
    showZapier,
  };
}

export default useServiceDefinitions;
