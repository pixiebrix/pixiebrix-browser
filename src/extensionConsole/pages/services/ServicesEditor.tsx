/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import styles from "@/extensionConsole/pages/services/PrivateServicesCard.module.scss";

import React, { useCallback, useContext, useState } from "react";
import { connect } from "react-redux";
import servicesSlice from "@/store/servicesSlice";
import Page from "@/layout/Page";
import { Card, Col, Row } from "react-bootstrap";
import { push } from "connected-react-router";
import ServiceEditorModal from "./ServiceEditorModal";
import PrivateServicesCard from "./PrivateServicesCard";
import ConnectExtensionCard from "./ConnectExtensionCard";
import { faCloud, faPlus } from "@fortawesome/free-solid-svg-icons";
import useServiceDefinitions from "./useServiceDefinitions";
import { services } from "@/background/messenger/api";
import ZapierModal from "@/extensionConsole/pages/services/ZapierModal";
import notify from "@/utils/notify";
import { useParams } from "react-router";
import BrickModal from "@/components/brickModalNoTags/BrickModal";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { uuidv4 } from "@/types/helpers";
import useAuthorizationGrantFlow from "@/hooks/useAuthorizationGrantFlow";
import { reportEvent } from "@/telemetry/events";
import { type Integration, type IntegrationConfig } from "@/types/serviceTypes";
import { type UUID } from "@/types/stringTypes";
import ReduxPersistenceContext from "@/store/ReduxPersistenceContext";

const { updateServiceConfig, deleteServiceConfig } = servicesSlice.actions;

type OwnProps = {
  updateServiceConfig: typeof updateServiceConfig;
  deleteServiceConfig: typeof deleteServiceConfig;
  navigate: typeof push;
};

const ServicesEditor: React.FunctionComponent<OwnProps> = ({
  updateServiceConfig,
  deleteServiceConfig,
  navigate,
}) => {
  const { id: configurationId } = useParams<{ id: UUID }>();
  // Newly created integration (to ensure it's visible in the table)
  const [newIntegration, setNewIntegration] = useState<Integration | null>(
    null
  );
  const [newConfigurationService, setNewConfigurationService] =
    useState<Integration>(null);
  const [newConfiguration, setNewConfiguration] =
    useState<IntegrationConfig>(null);
  const { flush: flushReduxPersistence } = useContext(ReduxPersistenceContext);

  const {
    activeConfiguration,
    serviceDefinitions,
    activeService,
    showZapier,
    isPending: servicesPending,
  } = useServiceDefinitions();

  const isConfiguring =
    configurationId &&
    ((newConfigurationService && newConfiguration) ||
      (activeService && activeConfiguration));

  const handleSave = useCallback(
    async (config) => {
      updateServiceConfig(config);
      notify.success(
        `${
          newConfigurationService ? "Created" : "Updated"
        } private configuration for ${
          (activeService ?? newConfigurationService)?.name
        }.`
      );

      setNewConfiguration(null);
      setNewConfigurationService(null);
      setNewIntegration(config as Integration);
      await flushReduxPersistence();

      try {
        await services.refresh();
      } catch (error) {
        notify.error({
          message:
            "Error refreshing service configurations, restart the PixieBrix extension",
          error,
        });
      }

      navigate("/services");
    },
    [
      updateServiceConfig,
      newConfigurationService,
      activeService,
      flushReduxPersistence,
      navigate,
    ]
  );

  const launchAuthorizationGrantFlow = useAuthorizationGrantFlow();

  const handleCreate = useCallback(
    async (service: Integration) => {
      reportEvent("ServiceAdd", {
        serviceId: service.id,
      });

      const definition = (serviceDefinitions ?? []).find(
        (x) => x.id === service.id
      );

      if (definition.isAuthorizationGrant) {
        void launchAuthorizationGrantFlow(service, { target: "_self" });
        return;
      }

      const config = {
        id: uuidv4(),
        label: undefined,
        serviceId: service.id,
        config: {},
      } as IntegrationConfig;

      setNewConfiguration(config);
      setNewConfigurationService(definition);
      navigate(`/services/${encodeURIComponent(config.id)}`);
    },
    [
      navigate,
      launchAuthorizationGrantFlow,
      serviceDefinitions,
      setNewConfiguration,
      setNewConfigurationService,
    ]
  );

  const handleDelete = useCallback(
    async (id) => {
      deleteServiceConfig({ id });
      notify.success(
        `Deleted private configuration for ${activeService?.name}`
      );
      navigate("/services/");

      await flushReduxPersistence();

      try {
        await services.refresh();
      } catch (error) {
        notify.error({
          message:
            "Error refreshing service configurations, restart the PixieBrix extension",
          error,
        });
      }
    },
    [deleteServiceConfig, activeService?.name, navigate, flushReduxPersistence]
  );

  return (
    <Page
      icon={faCloud}
      title="Integrations"
      description="Configure external accounts, resources, and APIs. Personal integrations are
          stored in your browser; they are never transmitted to the PixieBrix servers or shared with your team"
      isPending={servicesPending}
      toolbar={
        <BrickModal
          onSelect={handleCreate}
          bricks={serviceDefinitions}
          modalClassName={styles.ModalOverride}
          selectCaption={
            <span>
              <FontAwesomeIcon icon={faPlus} className="mr-1" /> Configure
            </span>
          }
          caption={
            <span>
              <FontAwesomeIcon icon={faPlus} />
              &nbsp;Add Integration
            </span>
          }
        />
      }
    >
      {showZapier && (
        <ZapierModal
          onClose={() => {
            navigate("/services");
          }}
        />
      )}
      {isConfiguring && (
        <ServiceEditorModal
          configuration={activeConfiguration ?? newConfiguration}
          service={activeService ?? newConfigurationService}
          onDelete={activeConfiguration && handleDelete}
          onClose={() => {
            navigate("/services");
          }}
          onSave={handleSave}
        />
      )}
      <Row>
        <Col>
          <ConnectExtensionCard />
        </Col>
      </Row>
      <Row>
        <Col>
          <Card>
            <Card.Header>Personal Integrations</Card.Header>
            <PrivateServicesCard
              navigate={navigate}
              services={serviceDefinitions}
              initialService={newIntegration}
            />
          </Card>
        </Col>
      </Row>
    </Page>
  );
};

export default connect(null, {
  updateServiceConfig,
  deleteServiceConfig,
  navigate: push,
})(ServicesEditor);
