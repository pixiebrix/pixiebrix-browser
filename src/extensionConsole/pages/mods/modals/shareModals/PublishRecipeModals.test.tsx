/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { render, screen } from "@/extensionConsole/testHelpers";
import PublishRecipeModals from "./PublishRecipeModals";
import { authSlice } from "@/auth/authSlice";
import { modModalsSlice } from "@/extensionConsole/pages/mods/modals/modModalsSlice";
import { type ModDefinition } from "@/types/modDefinitionTypes";
import { type AuthState } from "@/auth/authTypes";
import { validateRegistryId } from "@/types/helpers";
import { useGetMarketplaceListingsQuery } from "@/services/api";
import { waitForEffect } from "@/testUtils/testHelpers";
import { appApiMock } from "@/testUtils/appApiMock";
import {
  defaultModDefinitionFactory,
  metadataFactory,
} from "@/testUtils/factories/modDefinitionFactories";
import { authStateFactory } from "@/testUtils/factories/authFactories";

let blueprint: ModDefinition;
let auth: AuthState;

jest.mock("@/modDefinitions/modDefinitionHooks", () => ({
  useOptionalModDefinition: jest.fn().mockImplementation(() => ({
    data: blueprint,
    isFetching: false,
  })),
}));

/**
 * Wrapper component to fetch marketplace listings because the PublishRecipeModals component does not fetch.
 */
const MarketplaceListingsWrapper: React.FC = ({ children }) => {
  useGetMarketplaceListingsQuery();
  return <>{children}</>;
};

beforeEach(() => {
  auth = authStateFactory();
  blueprint = defaultModDefinitionFactory({
    metadata: metadataFactory({
      id: validateRegistryId(`${auth.scope}/test`),
    }),
  });

  // XXX: why do we need to call this if it's already defined in appApiMock?
  appApiMock.onGet("/api/marketplace/listings/").reply(200, []);
});

afterEach(() => {
  jest.clearAllMocks();
});

test("renders publish modal", async () => {
  render(
    <MarketplaceListingsWrapper>
      <PublishRecipeModals />
    </MarketplaceListingsWrapper>,
    {
      setupRedux(dispatch) {
        dispatch(authSlice.actions.setAuth(auth));

        dispatch(
          modModalsSlice.actions.setPublishContext({
            blueprintId: blueprint.metadata.id,
          })
        );
      },
    }
  );

  await waitForEffect();

  expect(screen.getByRole("dialog")).toMatchSnapshot();
});

test("renders edit publish modal", async () => {
  blueprint.sharing.public = true;

  render(
    <MarketplaceListingsWrapper>
      <PublishRecipeModals />
    </MarketplaceListingsWrapper>,
    {
      setupRedux(dispatch) {
        dispatch(authSlice.actions.setAuth(auth));

        dispatch(
          modModalsSlice.actions.setPublishContext({
            blueprintId: blueprint.metadata.id,
          })
        );
      },
    }
  );

  await waitForEffect();

  expect(screen.getByRole("dialog")).toMatchSnapshot();
});

test("renders cancel publish modal", async () => {
  blueprint.sharing.public = true;

  render(
    <MarketplaceListingsWrapper>
      <PublishRecipeModals />
    </MarketplaceListingsWrapper>,
    {
      setupRedux(dispatch) {
        dispatch(authSlice.actions.setAuth(auth));

        dispatch(
          modModalsSlice.actions.setPublishContext({
            blueprintId: blueprint.metadata.id,
            cancelingPublish: true,
          })
        );
      },
    }
  );

  await waitForEffect();

  expect(screen.getByRole("dialog")).toMatchSnapshot();
});
