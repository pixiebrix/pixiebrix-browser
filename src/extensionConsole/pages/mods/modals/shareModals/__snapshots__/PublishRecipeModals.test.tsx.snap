// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders cancel publish modal 1`] = `
<div
  aria-modal="true"
  class="fade modal show"
  role="dialog"
  style="display: block;"
  tabindex="-1"
>
  <div
    class="modal-dialog"
  >
    <div
      class="modal-content"
    >
      <div
        class="modal-body"
      >
        <div
          class="root"
          data-testid="loader"
        >
          <span
            style="width: 57px; font-size: 0px; display: inline-block;"
          >
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 1.06444705606819s 0.26444705606819013s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 1.3487942280284695s 0.5487942280284694s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 0.9424787069565245s 0.14247870695652448s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 0.6040398729413513s -0.19596012705864868s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 1.40489940599664s 0.6048994059966399s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 0.9841883751436941s 0.18418837514369407s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 0.9794749739806473s 0.17947497398064732s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 1.57230410370597s 0.7723041037059699s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 1.0586418776970006s 0.2586418776970007s infinite ease;"
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders edit publish modal 1`] = `
<div
  aria-modal="true"
  class="fade modal show"
  role="dialog"
  style="display: block;"
  tabindex="-1"
>
  <div
    class="modal-dialog"
  >
    <div
      class="modal-content"
    >
      <div
        class="modal-body"
      >
        <div
          class="root"
          data-testid="loader"
        >
          <span
            style="width: 57px; font-size: 0px; display: inline-block;"
          >
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 1.3392299487066919s 0.5392299487066918s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 1.1334504568999262s 0.3334504568999263s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 1.2151793963667465s 0.41517939636674644s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 0.985847862421927s 0.18584786242192702s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 1.376280072053373s 0.576280072053373s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 0.9004810700961631s 0.10048107009616308s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 0.973598794428227s 0.17359879442822707s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 0.9390783395255372s 0.13907833952553722s infinite ease;"
            />
            <span
              style="display: inline-block; background-color: rgb(0, 0, 0); width: 15px; height: 15px; margin: 2px; border-radius: 100%; animation-fill-mode: both; animation: react-spinners-GridLoader-grid 1.2735336356892586s 0.4735336356892587s infinite ease;"
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders publish modal 1`] = `
<div
  aria-modal="true"
  class="fade modal show"
  role="dialog"
  style="display: block;"
  tabindex="-1"
>
  <div
    class="modal-dialog"
  >
    <div
      class="modal-content"
    >
      <div
        class="modal-header"
      >
        <div
          class="modal-title h4"
        >
          Publish to Marketplace
        </div>
        <button
          class="close"
          type="button"
        >
          <span
            aria-hidden="true"
          >
            Ã—
          </span>
          <span
            class="sr-only"
          >
            Close
          </span>
        </button>
      </div>
      <div
        class="modal-body"
      >
        <p>
          On Submit, the public link to this mod will be shared with the
           
          <a
            href="https://www.pixiebrix.com/marketplace/"
            rel="noreferrer noopener"
            target="blank"
          >
            PixieBrix Marketplace
          </a>
           
          admin team, who will review your submission and publish your mod.
        </p>
        <p>
          As soon as you Submit, the public link below will work for anyone, so you can start sharing right away!
        </p>
        <p
          class="mb-1"
        >
          Public link to share:
        </p>
        <div
          class="input-group"
        >
          <input
            class="form-control"
            readonly=""
            type="text"
            value="https://app.pixiebrix.com/activate?id=@user1/test"
          />
          <div
            class="input-group-append"
          >
            <button
              class="btn btn-info"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-copy fa-w-14 "
                data-icon="copy"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 0 0 358.059 0H352v96h96v-6.059a24 24 0 0 0-7.029-16.97z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="modal-footer"
      >
        <button
          class="btn btn-link"
          type="button"
        >
          Cancel
        </button>
        <button
          class="btn btn-primary"
          type="button"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</div>
`;
